ksql-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryRoutingFunctionalTest.java

```diff
@@ -0,0 +1,398 @@
+      .withProperty(KsqlRestConfig.LISTENERS_CONFIG, "http://localhost:8087")
+      .withProperty(KsqlRestConfig.ADVERTISED_LISTENER_CONFIG, "http://localhost:8087")
+      .withProperties(COMMON_CONFIG)
+      .build();
+
+  @ClassRule
+  public static final RuleChain CHAIN = RuleChain
+      .outerRule(Retry.of(3, ZooKeeperClientException.class, 3, TimeUnit.SECONDS))
+      .around(TEST_HARNESS);
+
+  @Rule
+  public final Timeout timeout = Timeout.builder()
+      .withTimeout(1, TimeUnit.MINUTES)
+      .withLookingForStuckThread(true)
+      .build();
+
+  @Rule
+  public final ExpectedException expectedException = ExpectedException.none();
+
+  @BeforeClass
+  public static void setUpClass() {
+    //Create topic with 1 partition to control who is active and standby
+    TEST_HARNESS.ensureTopics(1, USER_TOPIC);
+
+    final AtomicLong timestampSupplier = new AtomicLong(BASE_TIME);
+
+    TEST_HARNESS.produceRows(
+        USER_TOPIC,
+        USER_PROVIDER,
+        VALUE_FORMAT,
+        timestampSupplier::getAndIncrement
+    );
+  }
+
+  @Before
+  public void setUp() {
+    //Create stream
+    makeAdminRequest(
+        REST_APP_0,
+        "CREATE STREAM " + USERS_STREAM
+            + " (" + USER_PROVIDER.ksqlSchemaString() + ")"
+            + " WITH ("
+            + "   kafka_topic='" + USER_TOPIC + "', "
+            + "   key='" + USER_PROVIDER.key() + "', "
+            + "   value_format='" + VALUE_FORMAT + "'"
+            + ");"
+    );
+    //Create table
+    output = KsqlIdentifierTestUtil.uniqueIdentifierName();
+    sql = "SELECT * FROM " + output + " WHERE ROWKEY = '" + KEY + "';";
+    List<KsqlEntity> res = makeAdminRequestWithResponse(
+        REST_APP_0,
+        "CREATE TABLE " + output + " AS"
+            + " SELECT COUNT(1) AS COUNT FROM " + USERS_STREAM
+            + " GROUP BY " + USER_PROVIDER.key() + ";"
+    );
+    QUERY_ID = extractQueryId(res.get(0).toString());
+    QUERY_ID = QUERY_ID.substring(0, QUERY_ID.length()-1);
+    waitForTableRows();
+    waitForStreamsMetadataToInitialize(
+        REST_APP_0, ImmutableList.of(host0, host1, host2), QUERY_ID);
+  }
+
+  @After
+  public void cleanUp() {
+    REST_APP_0.closePersistentQueries();
+    REST_APP_0.dropSourcesExcept();
+  }
+
+  @AfterClass
+  public static void classTearDown() {
+    TMP.delete();
+  }
+
+  @Test
+  public void shouldQueryActiveWhenActiveAliveQueryIssuedToStandby() {
+    // Given:
+    ClusterFormation clusterFormation = findClusterFormation(REST_APP_0, REST_APP_1, REST_APP_2);
+    waitForClusterToBeDiscovered(clusterFormation.standBy.right, 3);
+    sendHeartbeartsForWindowLength(
+        clusterFormation.standBy.right, clusterFormation.active.left, 2000);
+    waitForRemoteServerToChangeStatus(
+        clusterFormation.standBy.right, clusterFormation.active.left, HighAvailabilityTestUtil::remoteServerIsUp);
+
+    // When:
+    final List<StreamedRow> rows_0 = makePullQueryRequest(clusterFormation.standBy.right, sql);
```

Comment:
IIUC since there is just 1 partition, a query to standby returning should mean it was answered by active? Technically, it can be answered by standby locally as well right? Let's assume we introduce a bug that we query locally first always before proxying.. Would this test still pass then?

Reply:
Correct, there is no way of checking which server serves a query. We could introduce this information in the response?

Reply:
yes.. how much work would that be? This would also be incredibly useful debug information

Reply:
This might be a breaking change? Not sure if any applications use the response and expect it to be of certain format. How much work? Not sure. Don't know that code path

Reply:
I was saying this a few days ago.  For the lag sorting functionality where you expect that either of two standbys could answer a query (and they're both alive and not too behind), a functional test for it couldn't know which was the one that answered.  It would be impossible to test that it was the lower lag host.

This would be a nice thing to add at some point.

Reply:
Adding a new response header won't be a breaking change .. its actually a good backwards compatible change... But I can see there is some non-trivial work here.. So lets file few another follow up.. For now, the tests definitely cover the happy path and if we have other tests that guard against (KsLocatorTest may be already has them) the bug I used as an example, we might be fine.

Reply:
Tracked here: #4455